### PR TITLE
Add support for merge path algorithm (csrmvEx) when csr_matrix multiply with a dense vector

### PR DESCRIPTION
**Pull request by wonghang**
Modify csr_matrix ``__mul__`` to use merge path algorithm by cuSPARSE csrmvEx (Issue: #2283)

The rationale behind using this algorithm instead can be found in the following slides:
http://images.nvidia.com/events/sc15/pdfs/sc15-Merge-Based-Parallel-Sparse-Matrix-Vector-Multiplication-merrill.pdf

The nvidia API is here:
https://docs.nvidia.com/cuda/cusparse/index.html#csrmvEx

All pointers must be aligned with 128 bytes. Therefore a checking is performed before using ``csrmvEx``, it will fallback to original ``csrmv`` if any one of the pointers is not aligned. I also found that if ``nnz == 0``, ``csrmvEx`` will generate an illegal memory access error while ``csrmv`` does not. Checking is also added and fallback to ``csrmv`` if so.

According to this, all pointers allocated by ``cudaMalloc`` should be aligned.
https://stackoverflow.com/questions/14082964/cuda-alignment-256bytes-seriously

### CODE REVIEW

File: cupy/cusparse.py [hunk 1]
```diff
@@ -103,6 +118,123 @@ def csrmv(a, x, y=None, alpha=1, beta=0, transa=False):
...
+    beta = _aligned_constant(beta, dtype).ctypes
+
+    assert a.data.data.ptr % 128 == 0
+    assert a.indptr.data.ptr % 128 == 0
+    assert a.indices.data.ptr % 128 == 0
+    assert x.data.ptr % 128 == 0
+    assert y.data.ptr % 128 == 0
+    assert alpha.data % 128 == 0
+    assert beta.data % 128 == 0
+
```
Comment (emcastillo) [inline cupy/cusparse.py]: just to be pedantic but could you use the `csrmvExIsAligned` function you defined above for this?

File: cupy/cusparse.py [hunk 2]
```diff
@@ -103,6 +118,123 @@ def csrmv(a, x, y=None, alpha=1, beta=0, transa=False):
...
+        y = cupy.zeros(m, dtype)
+
+    datatype = _dtype_to_DataType(dtype)
+    algmode = cusparse.CUSPARSE_ALG_MERGE_PATH if \
+        merge_path else cusparse.CUSPARSE_ALG_NAIVE
+    transa_flag = cusparse.CUSPARSE_OPERATION_NON_TRANSPOSE
+
+    alpha = _aligned_constant(alpha, dtype).ctypes
+    beta = _aligned_constant(beta, dtype).ctypes
+
```
Comment (emcastillo) [inline cupy/cusparse.py]: According to cusparse documentation examples, it doesnt seem that alpha and beta need to be aligned to 128 bytes although they are being passed as a pointer.

I will probably be wrong but could you double-check this please.

https://docs.nvidia.com/cuda/cusparse/index.html#unique_2009252207

**Comment by wonghang** [comment]
Sure.

Be honest, I worked to align alpha and beta too before I realized that
test_csr.py would send a matrix with nnz=0. I may also be wrong. Both of
them are possible reason of illegal memory access error.

Let’s me do more checking tomorrow.

「emcastillo <notifications@github.com>」在 2019年7月3日 週三，16:27 寫道：

File: cupy/cusparse.py (hunk 2 above)
Comment (wonghang) [reply cupy/cusparse.py]: Sure.

Be honest, I worked to align alpha and beta too before I realized that
test_csr.py would send a matrix with nnz=0. I may also be wrong. Both of
them are possible reason of illegal memory access error.

Let’s me do more checking tomorrow.

Comment (emcastillo) [reply cupy/cusparse.py]: Thanks!

Comment (wonghang) [reply cupy/cusparse.py]: You are right. I tested with the following:

```python
def _misaligned_constant(a,dtype,alignment=128,misalign=1):
    a = numpy.array(a, dtype)
    if a.ctypes.data % alignment != 0:
        return a

    b = numpy.empty((alignment+a.itemsize+misalign, ), 'b')
    offset = b.ctypes.data % alignment
    if offset > 0:
        idx = alignment - offset + misalign
    else:
        idx = misalign
    b = b[idx:idx+a.itemsize]
    aa = b.view(dtype)
    aa[0] = a
    assert aa.ctypes.data % alignment == misalign
    return aa
```

and use the following instead:

```python
    alpha = _misaligned_constant(alpha, dtype).ctypes
    beta = _misaligned_constant(beta, dtype).ctypes

```

csrmvEx looks good.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins failure is unrelated.

**Comment by wonghang** [comment]
Hi, sometime when I am using this feature (run for a sufficient long time), I will get some mystery error and I do not have a deterministic way to reproduce it. I use it in an iterative solver.

```
  File "/usr/local/lib/python3.5/dist-packages/cupy-7.0.0b1-py3.5-linux-x86_64.egg/cupyx/scipy/sparse/csr.py", line 121, in __mul__
    return cusparse.csrmvEx(self, other)
  File "/usr/local/lib/python3.5/dist-packages/cupy-7.0.0b1-py3.5-linux-x86_64.egg/cupy/cusparse.py", line 199, in csrmvEx
    y = cupy.zeros(m, dtype)
  File "/usr/local/lib/python3.5/dist-packages/cupy-7.0.0b1-py3.5-linux-x86_64.egg/cupy/creation/basic.py", line 180, in zeros
    a.data.memset_async(0, a.nbytes)
  File "cupy/cuda/memory.pyx", line 470, in cupy.cuda.memory.MemoryPointer.memset_async
  File "cupy/cuda/memory.pyx", line 485, in cupy.cuda.memory.MemoryPointer.memset_async
  File "cupy/cuda/runtime.pyx", line 317, in cupy.cuda.runtime.memsetAsync
  File "cupy/cuda/runtime.pyx", line 147, in cupy.cuda.runtime.check_status
cupy.cuda.runtime.CUDARuntimeError: cudaErrorInvalidValue: invalid argument
```

Just want to tell your team that csrmvEx is not that stable. You may want to remove it in the release (stable version)

(Maybe related #2005 but I did check the result from ``to_csr()`` is correct in my case)

**Pull request merged**